File: backend-php/models/Sale.php
<?php
/**
 * Sale Model
 * Handle sales operations
 */

require_once __DIR__ . '/../config/Database.php';

class Sale {
    
    private $db;
    
    public function __construct() {
        $this->db = Database::getInstance();
    }
    
    public function getAll() {
        $sql = "SELECT s.*, c.CustomerName
                FROM Sales s
                LEFT JOIN Customers c ON s.CustomerID = c.CustomerID
                ORDER BY s.SaleID DESC";
        
        $stmt = $this->db->query($sql);
        $sales = $this->db->fetchAll($stmt);
        
        // Get items for each sale
        foreach ($sales as &$sale) {
            $sale['items'] = $this->getSaleDetails($sale['SaleID']);
        }
        
        return $sales;
    }
    
    public function getById($id) {
        $sql = "SELECT s.*, c.CustomerName
                FROM Sales s
                LEFT JOIN Customers c ON s.CustomerID = c.CustomerID
                WHERE s.SaleID = ?";
        
        $stmt = $this->db->query($sql, [$id]);
        $sale = $this->db->fetchOne($stmt);
        
        if ($sale) {
            $sale['items'] = $this->getSaleDetails($id);
        }
        
        return $sale;
    }
    
    private function getSaleDetails($saleId) {
        $sql = "SELECT sd.*, p.ProductName, p.Barcode
                FROM SaleDetails sd
                JOIN Products p ON sd.ProductID = p.ProductID
                WHERE sd.SaleID = ?";
        
        $stmt = $this->db->query($sql, [$saleId]);
        return $this->db->fetchAll($stmt);
    }
    
    public function create($data) {
        try {
            $this->db->beginTransaction();
            
            // Calculate totals
            $totalAmount = 0;
            foreach ($data['items'] as $item) {
                $totalAmount += $item['Quantity'] * $item['UnitPrice'];
            }
            
            $discount = $data['Discount'] ?? 0;
            $tax = $data['Tax'] ?? 0;
            $finalAmount = $totalAmount - $discount + $tax;
            
            // Generate invoice number
            $countSql = "SELECT COUNT(*) as count FROM Sales";
            $countStmt = $this->db->query($countSql);
            $countRow = $this->db->fetchOne($countStmt);
            $count = $countRow['count'] + 1;
            $invoiceNumber = 'INV-' . date('Ymd') . '-' . str_pad($count, 4, '0', STR_PAD_LEFT);
            
            // Create sale
            $sql = "INSERT INTO Sales (InvoiceNumber, CustomerID, SaleDate, TotalAmount, Discount, Tax, FinalAmount, PaymentMethod, Status, Notes) 
                    VALUES (?, ?, GETDATE(), ?, ?, ?, ?, ?, ?, ?)";
            
            $params = [
                $invoiceNumber,
                $data['CustomerID'] ?? null,
                $totalAmount,
                $discount,
                $tax,
                $finalAmount,
                $data['PaymentMethod'],
                'مكتملة',
                $data['Notes'] ?? null
            ];
            
            $this->db->execute($sql, $params);
            $saleId = $this->db->lastInsertId();
            
            // Add sale details
            foreach ($data['items'] as $item) {
                $totalPrice = $item['Quantity'] * $item['UnitPrice'];
                
                $detailSql = "INSERT INTO SaleDetails (SaleID, ProductID, Quantity, UnitPrice, TotalPrice) 
                              VALUES (?, ?, ?, ?, ?)";
                
                $this->db->execute($detailSql, [
                    $saleId,
                    $item['ProductID'],
                    $item['Quantity'],
                    $item['UnitPrice'],
                    $totalPrice
                ]);
                
                // Update stock
                $stockSql = "UPDATE Products SET Stock = Stock - ? WHERE ProductID = ?";
                $this->db->execute($stockSql, [$item['Quantity'], $item['ProductID']]);
            }
            
            // Update customer balance
            if (!empty($data['CustomerID'])) {
                $balanceSql = "UPDATE Customers SET Balance = Balance + ? WHERE CustomerID = ?";
                $this->db->execute($balanceSql, [$finalAmount, $data['CustomerID']]);
            }
            
            $this->db->commit();
            
            return $this->getById($saleId);
            
        } catch (Exception $e) {
            $this->db->rollback();
            throw $e;
        }
    }
    
    public function getDailyReport($date = null) {
        $date = $date ?? date('Y-m-d');
        
        $sql = "SELECT s.*, c.CustomerName
                FROM Sales s
                LEFT JOIN Customers c ON s.CustomerID = c.CustomerID
                WHERE CAST(s.SaleDate AS DATE) = ?
                ORDER BY s.SaleID DESC";
        
        $stmt = $this->db->query($sql, [$date]);
        $sales = $this->db->fetchAll($stmt);
        
        $totalSales = 0;
        $totalDiscount = 0;
        $totalTax = 0;
        
        foreach ($sales as &$sale) {
            $sale['items'] = $this->getSaleDetails($sale['SaleID']);
            $totalSales += $sale['FinalAmount'];
            $totalDiscount += $sale['Discount'];
            $totalTax += $sale['Tax'];
        }
        
        return [
            'sales' => $sales,
            'summary' => [
                'totalSales' => $totalSales,
                'totalDiscount' => $totalDiscount,
                'totalTax' => $totalTax,
                'count' => count($sales),
                'date' => $date
            ]
        ];
    }
}

File: lib/models/sale.dart
class Sale {
  final int? id;
  final String invoiceNumber;
  final int? customerId;
  final String? customerName;
  final double totalAmount;
  final double discount;
  final double tax;
  final double finalAmount;
  final double paidAmount;
  final double remainingAmount;
  final String paymentMethod; // cash, credit, bank_transfer
  final String status; // completed, pending, cancelled
  final String? notes;
  final DateTime createdAt;
  final DateTime? updatedAt;
  final List<SaleItem> items;

  Sale({
    this.id,
    required this.invoiceNumber,
    this.customerId,
    this.customerName,
    required this.totalAmount,
    this.discount = 0.0,
    this.tax = 0.0,
    required this.finalAmount,
    required this.paidAmount,
    required this.remainingAmount,
    this.paymentMethod = 'cash',
    this.status = 'completed',
    this.notes,
    DateTime? createdAt,
    this.updatedAt,
    this.items = const [],
  }) : createdAt = createdAt ?? DateTime.now();

  Map<String, dynamic> toMap() {
    return {
      'id': id,
      'invoice_number': invoiceNumber,
      'customer_id': customerId,
      'customer_name': customerName,
      'total_amount': totalAmount,
      'discount': discount,
      'tax': tax,
      'final_amount': finalAmount,
      'paid_amount': paidAmount,
      'remaining_amount': remainingAmount,
      'payment_method': paymentMethod,
      'status': status,
      'notes': notes,
      'created_at': createdAt.toIso8601String(),
      'updated_at': updatedAt?.toIso8601String(),
    };
  }

  factory Sale.fromMap(Map<String, dynamic> map) {
    // تحويل items من JSON إلى List<SaleItem>
    List<SaleItem> items = [];
    if (map['items'] != null && map['items'] is List) {
      items = (map['items'] as List)
          .map((item) => SaleItem.fromMap(item as Map<String, dynamic>))
          .toList();
    }

    return Sale(
      id: map['id'] ?? map['SaleID'],
      invoiceNumber: map['invoice_number'] ?? map['InvoiceNumber'] ?? '',
      customerId: map['customer_id'] ?? map['CustomerID'],
      customerName: map['customer_name'] ?? map['CustomerName'],
      totalAmount: (map['total_amount'] ?? map['TotalAmount'] ?? 0).toDouble(),
      discount: (map['discount'] ?? map['Discount'] ?? 0).toDouble(),
      tax: (map['tax'] ?? map['Tax'] ?? 0).toDouble(),
      finalAmount: (map['final_amount'] ?? map['FinalAmount'] ?? 0).toDouble(),
      paidAmount: (map['paid_amount'] ?? map['PaidAmount'] ?? 0).toDouble(),
      remainingAmount:
          (map['remaining_amount'] ?? map['RemainingAmount'] ?? 0).toDouble(),
      paymentMethod: map['payment_method'] ?? map['PaymentMethod'] ?? 'cash',
      status: map['status'] ?? map['Status'] ?? 'completed',
      notes: map['notes'] ?? map['Notes'],
      createdAt: map['created_at'] != null
          ? DateTime.parse(map['created_at'])
          : (map['CreatedAt'] != null
              ? DateTime.parse(map['CreatedAt'])
              : DateTime.now()),
      updatedAt: map['updated_at'] != null
          ? DateTime.parse(map['updated_at'])
          : (map['UpdatedAt'] != null
              ? DateTime.parse(map['UpdatedAt'])
              : null),
      items: items,
    );
  }

  Sale copyWith({
    int? id,
    String? invoiceNumber,
    int? customerId,
    String? customerName,
    double? totalAmount,
    double? discount,
    double? tax,
    double? finalAmount,
    double? paidAmount,
    double? remainingAmount,
    String? paymentMethod,
    String? status,
    String? notes,
    DateTime? createdAt,
    DateTime? updatedAt,
    List<SaleItem>? items,
  }) {
    return Sale(
      id: id ?? this.id,
      invoiceNumber: invoiceNumber ?? this.invoiceNumber,
      customerId: customerId ?? this.customerId,
      customerName: customerName ?? this.customerName,
      totalAmount: totalAmount ?? this.totalAmount,
      discount: discount ?? this.discount,
      tax: tax ?? this.tax,
      finalAmount: finalAmount ?? this.finalAmount,
      paidAmount: paidAmount ?? this.paidAmount,
      remainingAmount: remainingAmount ?? this.remainingAmount,
      paymentMethod: paymentMethod ?? this.paymentMethod,
      status: status ?? this.status,
      notes: notes ?? this.notes,
      createdAt: createdAt ?? this.createdAt,
      updatedAt: updatedAt ?? this.updatedAt,
      items: items ?? this.items,
    );
  }
}

class SaleItem {
  final int? id;
  final int? saleId;
  final int productId;
  final String productName;
  final String productBarcode;
  final int quantity;
  final double unitPrice;
  final double totalPrice;
  final double? discount;

  SaleItem({
    this.id,
    this.saleId,
    required this.productId,
    required this.productName,
    required this.productBarcode,
    required this.quantity,
    required this.unitPrice,
    required this.totalPrice,
    this.discount,
  });

  Map<String, dynamic> toMap() {
    return {
      'id': id,
      'sale_id': saleId,
      'product_id': productId,
      'product_name': productName,
      'product_barcode': productBarcode,
      'quantity': quantity,
      'unit_price': unitPrice,
      'total_price': totalPrice,
      'discount': discount,
    };
  }

  factory SaleItem.fromMap(Map<String, dynamic> map) {
    return SaleItem(
      id: map['id'],
      saleId: map['sale_id'],
      productId: map['product_id'],
      productName: map['product_name'],
      productBarcode: map['product_barcode'],
      quantity: map['quantity'],
      unitPrice: map['unit_price'].toDouble(),
      totalPrice: map['total_price'].toDouble(),
      discount: map['discount']?.toDouble(),
    );
  }
}
